### DISCUSSION TITLE
hard boundary conditions and anchors

### DISCUSSION THREAD

**Discussion by pkpilly** [lululxvi/deepxde#886] [Q&A]
It is not clear how data (via icbc.PointSetBC and anchors) can be used for training the PINN when hard boundary conditions are enforced (via apply_output_transforms). Any guidance will be appreciated.

**Comment by praksharma** [comment]
Sampling and output transform are two separate things. Why is that a problem? Put your anchor points (manual PDE points) and BC points (manual BC points) as 

```python
data = dde.data.TimePDE(
    geomtime,
    pde,
    [bc, ic, observe_y],
    num_domain=40,
    num_boundary=20,
    num_initial=10,
    anchors=observe_x,
    solution=func,
    num_test=10000,
)
```
Hard constraining is just a function binded to the output neurons. Just use them from the tutorials. 

```python
net.apply_output_transform(FUNCTION)
```

Check [this](https://github.com/lululxvi/deepxde/blob/15078f65a18c77fc2c68cad6e7782647d895cc6f/deepxde/nn/pytorch/nn.py#L19). They are independent.
